substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrTypeRepository.java

```diff
@@ -25,74 +25,47 @@
+import com.oracle.svm.jfr.traceid.JfrTraceIdLoadBarrier;
 import org.graalvm.nativeimage.Platform;
 import org.graalvm.nativeimage.Platforms;
 
 import com.oracle.svm.core.annotate.Uninterruptible;
-import com.oracle.svm.core.annotate.UnknownObjectField;
-import com.oracle.svm.core.heap.Heap;
-import com.oracle.svm.core.hub.DynamicHub;
-import com.oracle.svm.core.locks.VMMutex;
 import com.oracle.svm.core.thread.VMOperation;
 
 public class JfrTypeRepository implements JfrRepository {
-    private final VMMutex mutex;
     private final JfrSymbolRepository symbolRepo;
-    @UnknownObjectField(types = boolean[].class) private boolean[] usedTypes;
-
-    private int count;
 
     @Platforms(Platform.HOSTED_ONLY.class)
     public JfrTypeRepository(JfrSymbolRepository symbolRepo) {
-        this.mutex = new VMMutex();
         this.symbolRepo = symbolRepo;
     }
 
-    @Platforms(Platform.HOSTED_ONLY.class)
-    public void initialize(int typeCount) {
-        usedTypes = new boolean[typeCount];
-    }
-
     @Uninterruptible(reason = "Epoch must not change while in this method.")
     public long getClassId(Class<?> clazz) {
-        int typeId = DynamicHub.fromClass(clazz).getTypeID();
-
-        if (!usedTypes[typeId]) {
-            mutex.lockNoTransition();
-            try {
-                if (!usedTypes[typeId]) {
-                    usedTypes[typeId] = true;
-                    count++;
-                }
-            } finally {
-                mutex.unlock();
-            }
-        }
-
-        return typeId;
+        return JfrTraceId.load(clazz);
     }
 
     @Override
     public void write(JfrChunkWriter writer) throws IOException {
         assert VMOperation.isInProgressAtSafepoint();
         writer.writeCompressedLong(JfrTypes.Class.getId());
-        writer.writeCompressedLong(count);
+        writer.writeCompressedLong(JfrTraceIdLoadBarrier.klassCount(true));
+
+        Consumer<Class<?>> kc = aClass -> writeKlass(aClass, writer);
+        JfrTraceIdLoadBarrier.doKlasses(kc, true);
+    }
 
-        // TODO: this won't work as the id doesn't match the order/index of the written data.
-        // JfrMethodRepository has a similar issue. Possible solutions are:
-        // - just write all klasses
-        // - use a long[] instead of the boolean[]
-        // - use an UninterruptibleHashmap
-        for (Class<?> clazz : Heap.getHeap().getClassList()) {
-            int id = DynamicHub.fromClass(clazz).getTypeID();
-            if (usedTypes[id]) {
-                usedTypes[id] = false;
-                writer.writeCompressedLong(0L); // classloader
-                writer.writeCompressedLong(symbolRepo.getSymbolId(clazz));
-                writer.writeCompressedLong(0); // package id
-                writer.writeCompressedLong(clazz.getModifiers());
-            }
+    private void writeKlass(Class<?> clazz, JfrChunkWriter writer) {
+        try {
+            writer.writeCompressedLong(0L); // classloader
```

Comment:
For what it's worth, I made the package data 'temporary' as Christian describes in our implementation; partly because of https://github.com/oracle/graal/issues/2989 being annoying to work-around, but also because it wasn't really needed to be held beforehand. It's likely doable for modules, classloaders, etc. as well, but I didn't verify at the time so left them as is. The overall performance difference should be minimal but should still be considered.

Reply:
Can you elaborate on the concern for allocations here?

With the epoch system writing the typeset should be done after the epoch switch safepoint and is one location where it is safe to be interrupted for someone else's safepoint.

Reply:
To be honest, not sure. Maybe I was a bit overreacting because I just made lots of trace-ID stuff uninterruptible. I need to think about it (and get away for a day or so from it...)

Reply:
I removed tracking of packages, modules and classloaders for now. It greatly simplifies the change. If we find that we need it, then we can still add it back.